python/mlx/nn/layers/recurrent.py

```diff
@@ -0,0 +1,542 @@
+# Copyright © 2023 Apple Inc.
+
+import math
+
+import mlx.core as mx
+from mlx.nn.layers import Dropout
+from mlx.nn.layers.base import Module
+
+
+class RecurrentCellBase(Module):
+    r"""Base class for Recurrent cells. Helper for shared
+    attributes and weights initialization.
+
+        Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state(s)
+        bias (bool): Whether to use biases or not
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.bias = bias
+
+    def _init_weights(self):
+        scale = 1.0 / math.sqrt(self.hidden_size)
+        Wx = mx.random.uniform(
+            low=-scale, high=scale, shape=(self.input_size, self.hidden_size)
+        )
+        Wh = self.Wxh = mx.random.uniform(
+            low=-scale, high=scale, shape=(self.hidden_size, self.hidden_size)
+        )
+
+        if self.bias:
+            bx = mx.random.uniform(low=-scale, high=scale, shape=(self.hidden_size,))
+            bh = mx.random.uniform(low=-scale, high=scale, shape=(self.hidden_size,))
+        else:
+            bx = bh = None
+
+        return Wx, Wh, bx, bh
+
+
+class RNNCell(RecurrentCellBase):
+    r"""Apply one step of Elman recurrent unit with tanh or relu
+    non-linearity.
+
+    Concretely:
+
+    .. math::
+
+        h' = \text{tanh} (W_{xh}x + b_{xh} + W_{hh}h + b_{hh})
+
+    The input :math:`x` has shape ``ND` or ``D`` and the hidden
+    state :math:`h` has shape ``NH`` or ``H``, where:
+        - ``N`` is the batch dimension
+        - ``D`` is the input's features dimension
+        - ``H`` is the hidden dimension
+
+    If `nonlinearity` is `relu`, then ReLU is used instead of tanh.
+
+    Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state :math:`h`
+        nonlinearity (str): Non-linearity to use. Valid values: `tanh`
+        or `relu`. Default: `tanh`.
+        bias (bool): Whether to use biases or not. Default: `True`.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        nonlinearity: str = "tanh",
+        bias: bool = True,
+    ):
+        super().__init__(input_size, hidden_size, bias)
+        self.input_size = input_size
+
+        self.Wxh, self.Whh, self.bh, self.bi = self._init_weights()
+
+        if nonlinearity == "tanh":
+            self.activation = mx.tanh
+        elif nonlinearity == "relu":
+            self.activation = lambda x: mx.maximum(x, 0)
+        else:
+            raise ValueError(
+                f"Unsupported nonlinearity `{nonlinearity}`. Use 'tanh' or 'relu'."
+            )
+
+    def __call__(self, x, h_prev):
+        out = x @ self.Wxh
+        out += h_prev @ self.Whh
```

Comment:
I think the problem with the cell model is that you are missing out on a big opportunity to batch. Namely you can do the whole sequence as a batched matmul with `self.Wxh` and then do the hidden state updates sequentially. 

I'm not sure if there is a clean way to do this with the RNN cell. 

Maybe it makes more sense to get rid  of the cell entirely and have an RNN layer class (gru layer, lstm layer, etc) and put the shared functionality in a helper methods?

Also I don't think you need to have those classes do multiple layers. I think PyTorch does it that way because cuDNN did it that way, but it's more inline with the library to have the user stack layers as they like.

Reply:
Good points, so I will get rid of the cell and make it into a single layer.